modules/cugraph/src/hypergraph.ts

```diff
@@ -0,0 +1,399 @@
+                                     categories,
+                                     edgeShape,
+                                     dropEdgeAttrs,
+                                     delim,
+                                     source,
+                                     target,
+                                     eventId,
+                                     category,
+                                     edgeType,
+                                     nodeType);
+  const events = new DataFrame({});
+  const nodes  = entities;
+
+  const graph = create_graph(edges, source, target);
+
+  return {nodes, edges, events, entities, graph};
+}
+
+function _compute_columns(values: DataFrame, columns: string[]|null, skip: string[]) {
+  const result: string[] = [];
+  for (const name of columns ?? values.names) {
+    if (!skip.includes(name)) { result.push(name); }
+  }
+  result.sort();
+  return result;
+}
+
+function _create_events(values: DataFrame,
+                        columns: string[],
+                        delim: string,
+                        dropNulls: boolean,
+                        eventId: string,
+                        nodeType: string) {
+  const series_map: {[key: string]: any} = {};
+  for (const name of columns) { series_map[name] = values.get(name); }
+
+  if (!(eventId in series_map)) {
+    series_map[eventId] =
+      Series.sequence({type: new Int32, init: 0, step: 1, size: values.numRows});
+  }
+
+  series_map[eventId]  = _prepend_str(series_map[eventId], eventId, delim);
+  series_map[nodeType] = _scalar_init('event', series_map[eventId].length);
+
+  if (!dropNulls) {
+    for (const name of columns) {
+      const col = series_map[name];
+      if (col instanceof StringSeries) { series_map[name] = col.replaceNulls('null'); }
+    }
+  }
+
+  return new DataFrame(series_map);
+}
+
+function _create_entity_nodes(events: DataFrame,
+                              columns: string[],
+                              dropNulls: boolean,
+                              categories: {[key: string]: string},
+                              delim: string,
+                              nodeId: string,
+                              category: string,
+                              nodeType: string) {
+  const node_dfs: DataFrame[] = [];
+
+  for (const name of columns) {
+    const cat = name in categories ? categories[name] : name;
+    let col   = events.get(name);
+    col       = col.unique();
+    col       = dropNulls ? col.dropNulls() : col;  // nansToNulls?
+    if (col.length == 0) { continue; }
+
+    const df = new DataFrame({
+      [name]: col,
+      [nodeId]: _prepend_str(col, cat, delim),
+      [category]: _scalar_init(cat, col.length),
+      [nodeType]: _scalar_init(name, col.length),
+    });
+    node_dfs.push(df);
+  }
+  const nodes = new DataFrame()
+                  .concat(...node_dfs)
+                  .dropDuplicates('first', true, true, [nodeId])  // correct defaults?
```

Comment:
I'm assuming still being determined?

Reply:
AFAICT these are correct, I will remove the comment 
```suggestion
                  .dropDuplicates('first', true, true, [nodeId])
```